### ISSUE TITLE
Refactoring ideas regarding `Sampler`

### ISSUE THREAD

**Issue by hyeok9855** [GFNOrg/torchgfn#195]
There are two key components, `GFlowNet` (in `src/gfn/gflownet`) and `Sampler` (in `src/gfn/samplers.py`).

I think `Sampler` is a redundant class, and we could get rid of it by incorporating the sampling logic inside of the `GFlowNet` class.

Specifically, [this line](https://github.com/GFNOrg/torchgfn/blob/13c059a2d01a6acb5b6fc89fe1ee4e76ad2d4d4d/src/gfn/gflownet/base.py#L102) defines a sampler object only to sample one batch of a trajectory, which can be replaced by a single method in `GFlowNet` class.

**Comment by josephdviviano** [comment]
Yes, @saleml and I have had this conversation a few times before. I don't see a major downside of the `Sampler` becoming a few methods in the base `GFlowNet` class without much issue. 

The only downside I can see is this: if a user needs to subclass the `Sampler` to implement a new method, it would still be possible with the logic integrated into the `GFlowNet`, but it would require a little more mental gymnastics. If they were simply subclassing the `Sampler`, adding some new features, and then passing it to their new `GFlowNet` the changes would be easier to read / understand.

I'm very open to integrating it.

**Comment by josephdviviano** [comment]
@hyeok9855 for this one - if you want to propose a PR that handles this merge I will be willing to review.

However I am wary of unnecessarily merging everything into a few giant classes. This is *exactly* the sort of thing that makes the library harder for other people to extend for their own purposes. What we want is a level of modularity so that it's easy for people to subclass and extend only the required components for a new gflownet method to work.

In this case, having our sampler be distinct from the gflownet is useless because there's only one kind of sampler. But I could imagine in the future having multiple kinds of samplers that are compatible with all of the gflownet modules. Does this make sense to you?

**Comment by hyeok9855** [comment]
Maybe I could implement the `LocalSearchSampler`, as my first meaningful contribution.

**Comment by josephdviviano** [comment]
I love this idea!

**Comment by hyeok9855** [comment]
#208

**Comment by hyeok9855** [comment]
**One thing to discuss regarding `sample_trajectories`**

Currently, we can sample trajectories in two different ways:

1. `GFlowNet.sample_trajectories`, which defines `Sampler` inside of the method (both in [PFBasedGFlowNet](https://github.com/GFNOrg/torchgfn/blob/6f132a869902584c06b7f8054e76dda696d78c94/src/gfn/gflownet/base.py#L101) and [FMGFlowNet](https://github.com/GFNOrg/torchgfn/blob/6f132a869902584c06b7f8054e76dda696d78c94/src/gfn/gflownet/flow_matching.py#L57))
2. `Sampler.sample_trajectories` outside of the `GFlowNet` obj. ([example](https://github.com/GFNOrg/torchgfn/blob/6f132a869902584c06b7f8054e76dda696d78c94/tutorials/examples/train_hypergrid_simple.py#L55))

I think this is redundant, so here are options to streamline:
1. Pass `Sampler` class as an argument of `GFlowNet`.
2. Remove `GFlowNet.sample_trajectories` and force to use `Sampler.sample_trajectories`.

Please share your ideas!

**Comment by josephdviviano** [comment]
This got overlooked, which is a shame because it's potentially API-breaking.

I think one solution would be to have the `gflownet.sample_trajectories` method pass `args*, kwargs*` to the internal sampler, and for the sampler itself to be optionally passed to the gflownet at construction.

I like this conceptually because the GFlowNet itself is an amortized sampler, so I would expect it to have sample method.

Thoughts? cc: @saleml @younik
